src/pj_param.c

```diff
@@ -1,20 +1,78 @@
 /* put parameters in linked list and retrieve */
-#include <projects.h>
 #include <stdio.h>
+#include <ctype.h>
 #include <string.h>
+#include "projects.h"
 
 /* create parameter list entry */
 paralist *pj_mkparam(char *str) {
-	paralist *newitem;
-
-	if((newitem = (paralist *)pj_malloc(sizeof(paralist) + strlen(str))) != NULL) {
-		newitem->used = 0;
-		newitem->next = 0;
-		if (*str == '+')
-			++str;
-		(void)strcpy(newitem->param, str);
-	}
-	return newitem;
+    paralist *newitem;
+
+    if((newitem = (paralist *)pj_malloc(sizeof(paralist) + strlen(str))) != NULL) {
+        newitem->used = 0;
+        newitem->next = 0;
+        if (*str == '+')
+            ++str;
+        (void)strcpy(newitem->param, str);
+    }
+    return newitem;
+}
+
+
+/* As pj_mkparam, but payload is terminated by whitespace, rather than 0 */
+paralist *pj_mkparam_ws (char *str) {
+    paralist *newitem;
+    size_t len = 0;
+
+    if (0==str)
+        return 0;
+
+    /* Find start and length of string */
+    while (isspace (*str))
+        str++;
+    while ((!isspace(str[len])) && 0!=str[len])
+        len++;
+    if (*str == '+') {
+        str++;
+        len--;
+    }
+
+    /* Use calloc to automagically 0-terminate the strncpy */
+    newitem = (paralist *) pj_calloc (1, sizeof(paralist) + len);
+    if (0==newitem)
+        return 0;
+    strncpy(newitem->param, str, len);
+
+    newitem->used = 0;
+    newitem->next = 0;
+
+    return newitem;
+}
+
+/**************************************************************************************/
+paralist *pj_param_exists (paralist *list, const char *parameter) {
+/***************************************************************************************
+    Determine whether a given parameter exists in a paralist. If it does, return
+    a pointer to the corresponding list element - otherwise return 0.
+
+    This function is equivalent to the pj_param (...) call with the "opt" argument
+    set to the parameter name preceeeded by a 't'. But by using this one, one avoids
+    writing the code allocating memory for a new copy of parameter name, and prepending
+    the t (for compile time known names, this is obviously not an issue).
+***************************************************************************************/
```

Comment:
Make an explicit statement about memory ownership and that there isn't a NUL terminator at the end of the param.

Reply:
Actually there is, so your comment is an indication that this is not described clearly enough. Will try to think up a better description.